de.metas.adempiere.adempiere/base/src/main/java/de/metas/bpartner/service/impl/BPartnerDAO.java

```diff
@@ -1057,16 +1056,23 @@ public BPartnerLocationId retrieveBPartnerLocationId(@NonNull final BPartnerLoca
 				.addOnlyActiveRecordsFilter();
 
 		queryBuilder.orderBy()
-				.addColumn(I_C_BP_Relation.COLUMNNAME_C_BP_Relation_ID);
+		.addColumn(I_C_BP_Relation.COLUMNNAME_C_BP_Relation_ID);
 
-		final I_C_BP_Relation relation = bpRelationQueryBuilder
+		final Optional<BPartnerLocationId> relBPLocationId = bpRelationQueryBuilder
 				.create()
-				.firstOnly(I_C_BP_Relation.class); // just added an UC
-		if (relation != null)
+				.stream()
+				.map(bpRelationRecord -> ofRelationRecord(bpRelationRecord))
+				.filter(bpRelation -> Objects.equals(query.getRelationBPartnerLocationId(), bpRelation.getBplocationId()))
```

Comment:
i think this is also add odds with the javadoc. the javadoc, sais "If set, then prefer...". from that, i would expect it to return a different bpRelation'S locationId if the specified one isn't found..